File: lua/pay_kit/protocol/core/error_codes.lua
--[[
Canonical structured error codes for the Lua MPP server.

Mirrors `python/src/solana_mpp/_errors.py`.
Every server-side rejection raises through `raise(code, message)`
which throws an `error({code = code, message = message})` table the
HTTP boundary then translates into a JSON 402 body carrying `code`,
`error`, and `message`. The three fields are kept distinct for
backward compatibility:

  - `error`   short human description of the failure
  - `message` longer human-readable detail (matches `error` today)
  - `code`    machine-readable code from this module

The Lua codebase used to mix two error shapes: `error('string')` in
the PR A server core and `error({code = '...', message = '...'})` in
the PR B Solana stack. Standardizing on the table shape lets every
HTTP boundary (harness adapter, simple-server, Kong plugin) surface
the same code field without per-callsite glue.

Code set (matches Python and the cross-SDK fault matrix at audit v2):

  charge_request_mismatch     the credential's amount, currency, or
                              recipient does not match the route's
                              expected charge request
  challenge_route_mismatch    the credential's method, intent, or
                              realm does not match this server (cross
                              route replay)
  challenge_verification_failed  HMAC id mismatch or signature does
                                 not validate against the secret
  challenge_expired           the challenge's expires timestamp has
                              passed
  payment_invalid             the on-chain transaction shape failed
                              one of the verifier's structural checks
                              (mint, amount, ATA, memo, fee payer,
                              compute budget, etc.)
  wrong_network               the transaction's recent_blockhash
                              belongs to a different network than the
                              server is configured for
  signature_consumed          the replay store has already recorded
                              this signature for the configured key
                              prefix
]]

local M = {}

M.CHARGE_REQUEST_MISMATCH = 'charge_request_mismatch'
M.CHALLENGE_ROUTE_MISMATCH = 'challenge_route_mismatch'
M.CHALLENGE_VERIFICATION_FAILED = 'challenge_verification_failed'
M.CHALLENGE_EXPIRED = 'challenge_expired'
M.PAYMENT_INVALID = 'payment_invalid'
M.WRONG_NETWORK = 'wrong_network'
M.SIGNATURE_CONSUMED = 'signature_consumed'

-- Set of canonical codes; used by the response builder to assert that no
-- failure path falls through with a missing or unrecognized code. The lookup
-- table doubles as a runtime allowlist.
M.ALL = {
  [M.CHARGE_REQUEST_MISMATCH] = true,
  [M.CHALLENGE_ROUTE_MISMATCH] = true,
  [M.CHALLENGE_VERIFICATION_FAILED] = true,
  [M.CHALLENGE_EXPIRED] = true,
  [M.PAYMENT_INVALID] = true,
  [M.WRONG_NETWORK] = true,
  [M.SIGNATURE_CONSUMED] = true,
}

--- Raise a structured error with the given canonical code and message.
--- Raises a table-shaped error so callers that pcall the SDK pick up the
--- code through `err.code`; bare string consumers continue to work via
--- `tostring(err)` (Lua converts the table through its `__tostring`).
function M.raise(code, message)
  if M.ALL[code] == nil then
    -- Defensive: surface programming mistakes (typo in a callsite,
    -- forgotten code constant) immediately rather than letting an
    -- unknown code reach the response builder. The error itself is
    -- raised as a string so the test harness sees the developer
    -- mistake plainly.
    error('invalid error code: ' .. tostring(code))
  end
  error({ code = code, message = tostring(message or '') })
end

--- Project an error value (any of: table, string, raised lua error) into a
--- response-builder shape: `{code, error, message}`. Tables that already
--- carry `code` keep it; bare strings get the generic
--- CHALLENGE_VERIFICATION_FAILED fallback so the response is never
--- missing a code field.
function M.to_response(err)
  if type(err) == 'table' and err.code ~= nil and M.ALL[err.code] then
    return {
      code = err.code,
      error = err.message or '',
      message = err.message or '',
    }
  end
  local message
  if type(err) == 'table' and err.message ~= nil then
    message = err.message
  else
    message = tostring(err or '')
  end
  return {
    code = M.CHALLENGE_VERIFICATION_FAILED,
    error = message,
    message = message,
  }
end

return M

File: ruby/lib/pay_core/error_codes.rb
# frozen_string_literal: true

module PayCore
  # Canonical structured error codes (audit v2 L6 / P1 lock, mirrored across
  # Ruby, PHP, Lua, Rust, TypeScript, Go, Python).
  #
  # Every 402 response body emitted by the server SDK carries a `code` field
  # with one of these constants. The body also keeps the legacy `error` and
  # `message` fields so a polyglot client that pre-dates L6 still works.
  #
  # `canonical_code` maps an MPP/x402 error message or a legacy code to the
  # right L6 canonical code. Unknown failure classes fall back to
  # `payment_invalid` so a 402 response always carries a canonical code.
  module ErrorCodes
    # The credential's claimed charge does not match the route's expected
    # charge (amount, recipient, currency, method details).
    CODE_CHARGE_REQUEST_MISMATCH = "charge_request_mismatch"

    # The credential was issued for a different route than the one being
    # requested (different pinned fields: realm, intent, method).
    CODE_CHALLENGE_ROUTE_MISMATCH = "challenge_route_mismatch"

    # HMAC verification failed on the challenge id.
    CODE_CHALLENGE_VERIFICATION_FAILED = "challenge_verification_failed"

    # The challenge's `expires` is in the past.
    CODE_CHALLENGE_EXPIRED = "challenge_expired"

    # The credential payload is malformed or fails on-chain verification:
    # decode error, instruction allowlist violation, signature shape error.
    CODE_PAYMENT_INVALID = "payment_invalid"

    # The credential was signed against a different network than the one the
    # server is configured for.
    CODE_WRONG_NETWORK = "wrong_network"

    # The on-chain signature has already been used to settle a previous charge.
    CODE_SIGNATURE_CONSUMED = "signature_consumed"

    CANONICAL_CODES = [
      CODE_CHARGE_REQUEST_MISMATCH,
      CODE_CHALLENGE_ROUTE_MISMATCH,
      CODE_CHALLENGE_VERIFICATION_FAILED,
      CODE_CHALLENGE_EXPIRED,
      CODE_PAYMENT_INVALID,
      CODE_WRONG_NETWORK,
      CODE_SIGNATURE_CONSUMED
    ].freeze

    # Mapping from legacy or per-language internal codes to canonical codes.
    # Mirrors python/src/solana_mpp/_errors.py `_LEGACY_TO_CANONICAL`.
    LEGACY_TO_CANONICAL = {
      "challenge-expired" => CODE_CHALLENGE_EXPIRED,
      "challenge-mismatch" => CODE_CHALLENGE_VERIFICATION_FAILED,
      "signature-consumed" => CODE_SIGNATURE_CONSUMED,
      "wrong-network" => CODE_WRONG_NETWORK,
      "amount-mismatch" => CODE_CHARGE_REQUEST_MISMATCH,
      "recipient-mismatch" => CODE_CHARGE_REQUEST_MISMATCH,
      "splits-exceed-amount" => CODE_CHARGE_REQUEST_MISMATCH,
      "invalid-payload" => CODE_PAYMENT_INVALID,
      "invalid-payload-type" => CODE_PAYMENT_INVALID,
      "invalid-config" => CODE_PAYMENT_INVALID,
      "missing-signature" => CODE_PAYMENT_INVALID,
      "missing-transaction" => CODE_PAYMENT_INVALID,
      "transaction-failed" => CODE_PAYMENT_INVALID,
      "transaction-not-found" => CODE_PAYMENT_INVALID,
      "no-transfer" => CODE_PAYMENT_INVALID
    }.freeze

    # Substring patterns that classify an SDK error message into a canonical
    # code when no explicit code was set at raise time. Ordered; first match
    # wins. Mirrors harness/src/canonical-codes.ts and
    # rust/src/bin/harness_server.rs::classify_canonical_code.
    MESSAGE_PATTERNS = [
      [/already consumed/i, CODE_SIGNATURE_CONSUMED],
      # Solana RPC's own duplicate-signature reject text. Surfaces when
      # an idempotent-resubmit reaches the RPC's per-blockhash signature
      # uniqueness check before (or instead of) the local replay store -
      # the matrix's charge-idempotent-resubmit pins this.
      [/already been processed/i, CODE_SIGNATURE_CONSUMED],
      [/challenge verification failed/i, CODE_CHALLENGE_VERIFICATION_FAILED],
      [/challenge expired/i, CODE_CHALLENGE_EXPIRED],
      [/signed against localnet but the server expects/i, CODE_WRONG_NETWORK],
      [/network mismatch/i, CODE_WRONG_NETWORK],
      [/amount mismatch/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/currency mismatch/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/recipient mismatch/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/method details mismatch/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/split amounts exceed total/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/too many splits/i, CODE_CHARGE_REQUEST_MISMATCH],
      [/credential method does not match/i, CODE_CHALLENGE_ROUTE_MISMATCH],
      [/credential intent is not a charge/i, CODE_CHALLENGE_ROUTE_MISMATCH],
      [/credential realm does not match/i, CODE_CHALLENGE_ROUTE_MISMATCH],
      [/unexpected program instruction/i, CODE_CHARGE_REQUEST_MISMATCH]
    ].freeze

    # Return the canonical L6 code for a code or an error message.
    #
    # Resolution order:
    #   1. The string is already a canonical L6 code.
    #   2. The string is a legacy kebab-case code with a known mapping.
    #   3. The string matches a classified message pattern.
    #   4. Fallback: `payment_invalid`.
    def self.canonical_code(code_or_message)
      return CODE_PAYMENT_INVALID if code_or_message.nil? || code_or_message.empty?
      return code_or_message if CANONICAL_CODES.include?(code_or_message)
      mapped = LEGACY_TO_CANONICAL[code_or_message]
      return mapped if mapped

      MESSAGE_PATTERNS.each do |pattern, code|
        return code if code_or_message.match?(pattern)
      end
      CODE_PAYMENT_INVALID
    end
  end
end
